### PR TITLE
Use function pointers for reverse pinvokes

### LINKED ISSUE #21672: Use function pointers in IISHttpServer
Convert these https://github.com/dotnet/aspnetcore/blob/aaaaf572fdf11088ee0efd205ff6834e16a36804/src/Servers/IIS/IIS/src/Core/IISHttpServer.cs#L25-L29 to use the new [function pointers feature in C#](https://github.com/dotnet/csharplang/blob/master/proposals/function-pointers.md)

cc @jkotas

### PR DESCRIPTION
**Pull request by davidfowl**
- Mostly a mechanical change to use function pointers

Fixes https://github.com/dotnet/aspnetcore/issues/21672

### CODE REVIEW

File: src/Servers/IIS/IIS/src/Core/IISHttpServer.cs [hunk 1]
```diff
@@ -134,7 +136,8 @@ public void Dispose()
             _nativeApplication.Dispose();
         }
 
-        private static NativeMethods.REQUEST_NOTIFICATION_STATUS HandleRequest(IntPtr pInProcessHandler, IntPtr pvRequestContext)
+        [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvCdecl) })]
+        private static NativeMethods.REQUEST_NOTIFICATION_STATUS HandleRequest(nint pInProcessHandler, nint pvRequestContext)
```
Comment (jkotas) [inline src/Servers/IIS/IIS/src/Core/IISHttpServer.cs:140]: It is not an improvement to use `nint` for pointers. Use `IntPtr` or `void*.

File: src/Servers/IIS/IIS/src/NativeMethods.cs [hunk 2]
```diff
@@ -56,14 +49,14 @@ public enum REQUEST_NOTIFICATION_STATUS
...
-        private static extern int register_callbacks(NativeSafeHandle pInProcessApplication,
-            PFN_REQUEST_HANDLER requestCallback,
-            PFN_SHUTDOWN_HANDLER shutdownCallback,
-            PFN_DISCONNECT_HANDLER disconnectCallback,
-            PFN_ASYNC_COMPLETION asyncCallback,
-            PFN_REQUESTS_DRAINED_HANDLER requestsDrainedHandler,
-            IntPtr pvRequestContext,
-            IntPtr pvShutdownContext);
+        private unsafe static extern int register_callbacks(NativeSafeHandle pInProcessApplication,
+            delegate* unmanaged[Cdecl]<nint, nint, REQUEST_NOTIFICATION_STATUS> requestCallback,
```
Comment (jkotas) [inline src/Servers/IIS/IIS/src/NativeMethods.cs:53]: The original delegates had default unmanaged calling convention. Explicitly specifying `Cdecl` here is changing behavior. 

I think this should be just `delegate* unmanaged<nint, nint, REQUEST_NOTIFICATION_STATUS>`. Similar for all other places.

Comment (davidfowl) [reply src/Servers/IIS/IIS/src/NativeMethods.cs:53]: I was confused about this. Will change it back.

File: src/Servers/IIS/IIS/src/Core/IISHttpServer.cs [hunk 3]
```diff
@@ -163,7 +166,8 @@ private static NativeMethods.REQUEST_NOTIFICATION_STATUS HandleRequest(IntPtr pI
             }
         }
 
-        private static bool HandleShutdown(IntPtr pvRequestContext)
+        [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvCdecl) })]
+        private static bool HandleShutdown(nint pvRequestContext)
```
Comment (AaronRobinsonMSFT) [inline src/Servers/IIS/IIS/src/Core/IISHttpServer.cs]: I can't recall off the top of my head, but I believe `bool` is still non-blittable in this case so I would instead use `byte` for a C++ bool or `int` for a Win32 bool scenario.

**Review by jkotas — APPROVED** [review]
LGTM. Nice!

File: src/Servers/IIS/IIS/src/Core/IISHttpServer.cs (hunk 3 above)
Comment (jkotas) [reply src/Servers/IIS/IIS/src/Core/IISHttpServer.cs]: Good catch. You should see tests failing because of this. This should use `int`.

Comment (davidfowl) [reply src/Servers/IIS/IIS/src/Core/IISHttpServer.cs]: Done!

Comment (davidfowl) [reply src/Servers/IIS/IIS/src/Core/IISHttpServer.cs]: This should be an analyzer error or something.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**